# Tugas 1 Perolehan Informasi Lanjut Genap 2022 - Pyterrier

Pada tugas ini, digunakan library Python bernama Pyterrier yang merupakan versi Python dari [IR tool Terrier](http://terrier.org/). Installasi Pyterrier bisa dilihat pada [link dokumentasi Pyterrier](https://pyterrier.readthedocs.io/en/latest/). Anda juga dapat menggunakan file tutorial pyterrier di scele untuk mengerjakan tugas ini. Kerjakan dengan melengkapi #TODO pada setiap cell yang diberikan. Berikan dokumentasi yang jelas pada code Anda (comment pada line code yang diperlukan penjelasan). Disarankan menggunakan Google Colab untuk memudahkan pekerjaan Anda.

## Prerequisites

Instalasi Pyterrier

In [ ]:
# TODO
# Lakukan instalasi pyterrier
!pip install python-terrier

     |████████████████████████████████| 97 kB 2.1 MB/s 
     |████████████████████████████████| 1.1 MB 11.9 MB/s 
     |████████████████████████████████| 69 kB 6.3 MB/s 
     |████████████████████████████████| 255 kB 37.8 MB/s 
     |████████████████████████████████| 41 kB 198 kB/s 
     |████████████████████████████████| 596 kB 31.7 MB/s 
     |████████████████████████████████| 1.2 MB 17.1 MB/s 
     |████████████████████████████████| 72 kB 856 kB/s 
     |████████████████████████████████| 126 kB 36.7 MB/s 
     |████████████████████████████████| 294 kB 26.1 MB/s 
     |████████████████████████████████| 6.4 MB 30.9 MB/s 
     |████████████████████████████████| 291 kB 34.7 MB/s 
     |████████████████████████████████| 59 kB 5.5 MB/s 
     |████████████████████████████████| 45 kB 2.2 MB/s 
  Created wheel for python-terrier: filename=python_terrier-0.8.0-py3-none-any.whl size=103883 sha256=f91223d781fbc7a331ce269d05c9a7281f80f183d5a69b960c5f54ae891dd4c2
  Stored in directory: /root/.cac

## Init 

In [ ]:
# TODO
# Melakukan inisialisasi pyterrier
import pyterrier as pt
if not pt.started():
  pt.init()

terrier-assemblies 5.6 jar-with-dependencies not found, downloading to /root/.pyterrier...
Done
terrier-python-helper 0.0.6 jar not found, downloading to /root/.pyterrier...
Done


PyTerrier 0.8.0 has loaded Terrier 5.6 (built by craigmacdonald on 2021-09-17 13:27)



## Indexing

Akan dilakukan `indexing` pada file documents.csv. 

In [ ]:
# jika menggunaakan Google Colab, mount data dari google drive

from google.colab import files
from google.colab import drive

drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# TODO
# import dan tampilkan documents.csv dalam pandas dataframe

import pandas as pd
df = pd.read_csv('../documents.csv')
df.head(10)

docno                                               text
0  1781133  it was used in landing craft during world war ...
1  2426736  after rejecting an offer from cambridge univer...
2  2224122  mat zan coached kuala lumpur fa in 1999 and wo...
3   219642  a barcode is a machine readable optical label ...
4  1728654  since the subordination of the monarchy under ...
5  1889917  the tournament brought together club champions...
6  1518473  in january 1948 ballen acquired the interest o...
7  1459418  it is particularly notable for its association...
8  1102956  they are also known as the wamba wamba wemba w...
9   931408  they moved south to england to cannock staffor...

In [ ]:
# TODO
# Melakukan indexing pada file documents.csv (hint: gunakan IterDictIndexer)

iter_indexer = pt.IterDictIndexer("./index")
indexref = iter_indexer.index(df.to_dict(orient="records"))

In [ ]:
# TODO
# Tampilkan statistik dari index yang sudah dibuat
index = pt.IndexFactory.of(indexref)
print(index.getCollectionStatistics().toString())

Number of documents: 369721
Number of terms: 674552
Number of postings: 30552936
Number of fields: 1
Number of tokens: 41306796
Field names: [text]
Positions:   false



## Retrieval

Melakukan `querying` pada documents yang sudah di-index.


In [ ]:
# TODO
# import dan tampilkan query dari query.csv dalam pandas dataframe
import pandas as pd
df = pd.read_csv('/../query14.csv')
df.head(10)

qid                     query
0    6835               joan of arc
1   19342         chinese americans
2   18363  fox broadcasting company
3  226344           parramatta eels
4   37955                  duisburg
5   19145   oxford university press
6   14445                     yukon
7  605066   tenants purchase scheme
8    2228            cthulhu mythos
9    2613                 cape town

### BM25

In [ ]:
# TODO
# Melakukan querying menggunakan scoring function BM25 dengan 
# hanya mengambil top 10 documents untuk setiap query
# Tampilkan hasil search
retr_BM25 = pt.BatchRetrieve(index, controls = {"wmodel": "BM25"})

retr_BM25.setControl("wmodel", "BM25")
retr_BM25.setControls({"wmodel": "BM25"})

res_BM25 = retr_BM25.transform(df)

res_BM25

qid   docid    docno  rank      score                query
0       6835  175366  2380216     0  28.017621          joan of arc
1       6835   34385  2185113     1  26.806747          joan of arc
2       6835  253566  1561926     2  26.234903          joan of arc
3       6835  123783   734482     3  25.998385          joan of arc
4       6835  341480  1059708     4  25.445372          joan of arc
...      ...     ...      ...   ...        ...                  ...
19148  13690  125971  1244890   995   4.596766  university of tulsa
19149  13690   58510  1070070   996   4.594988  university of tulsa
19150  13690  202147   720978   997   4.594988  university of tulsa
19151  13690  331098  2134532   998   4.594988  university of tulsa
19152  13690    5601  1856133   999   4.593978  university of tulsa

[19153 rows x 6 columns]

**Menampilkan hasil search:**

Top 10 documents untuk setiap query

In [ ]:
res_BM25_TOP10 = (retr_BM25 % 10).transform(df)
res_BM25_TOP10

qid   docid    docno  rank      score                query
0       6835  175366  2380216     0  28.017621          joan of arc
1       6835   34385  2185113     1  26.806747          joan of arc
2       6835  253566  1561926     2  26.234903          joan of arc
3       6835  123783   734482     3  25.998385          joan of arc
4       6835  341480  1059708     4  25.445372          joan of arc
...      ...     ...      ...   ...        ...                  ...
18158  13690  238780    13690     5  21.113617  university of tulsa
18159  13690  171309  2166599     6  20.747642  university of tulsa
18160  13690   90797  1356524     7  20.266953  university of tulsa
18161  13690   85822  1032490     8  20.024378  university of tulsa
18162  13690  315002  2075501     9  19.783245  university of tulsa

[300 rows x 6 columns]

In [ ]:
# TODO
# Simpan hasil search query BM25 dengan penamaan file Tugas1_[NPM]_BM25.res
pt.io.write_results(res_BM25_TOP10, "../Tugas1_BM25.res")

### Dirichlet Language Model

In [ ]:
# TODO
# Melakukan querying menggunakan scoring function 
# Dirichlet Language Model dengan hanya mengambil top 10 documents untuk setiap query
# Tampilkan hasil search seluruh query
retr_DLM = pt.BatchRetrieve(index, controls = {"wmodel": "DirichletLM"})

retr_DLM.setControl("wmodel", "DirichletLM")
retr_DLM.setControls({"wmodel": "DirichletLM"})

res_DLM = retr_DLM.transform(df)

res_DLM

qid   docid    docno  rank      score                query
0       6835  175366  2380216     0  10.687178          joan of arc
1       6835  123783   734482     1  10.037802          joan of arc
2       6835   34385  2185113     2   9.896506          joan of arc
3       6835  253566  1561926     3   9.884301          joan of arc
4       6835  341480  1059708     4   9.866639          joan of arc
...      ...     ...      ...   ...        ...                  ...
19148  13690   84831   512129   995   1.028031  university of tulsa
19149  13690  106653   872315   996   1.028031  university of tulsa
19150  13690  107766  1891669   997   1.028031  university of tulsa
19151  13690  121958  2083199   998   1.028031  university of tulsa
19152  13690  152247   401094   999   1.028031  university of tulsa

[19153 rows x 6 columns]

Top 10 documents untuk setiap query:

In [ ]:
res_DLM_TOP10 = (retr_DLM % 10).transform(df)
res_DLM_TOP10

qid   docid    docno  rank      score                query
0       6835  175366  2380216     0  10.687178          joan of arc
1       6835  123783   734482     1  10.037802          joan of arc
2       6835   34385  2185113     2   9.896506          joan of arc
3       6835  253566  1561926     3   9.884301          joan of arc
4       6835  341480  1059708     4   9.866639          joan of arc
...      ...     ...      ...   ...        ...                  ...
18158  13690   56093  2062276     5   7.688806  university of tulsa
18159  13690  238780    13690     6   7.522458  university of tulsa
18160  13690  171309  2166599     7   7.509263  university of tulsa
18161  13690  291751   335604     8   7.490041  university of tulsa
18162  13690   83317  1043086     9   7.486192  university of tulsa

[300 rows x 6 columns]

In [ ]:
# TODO
# Simpan hasil search query Dirichlet Language Model dengan penamaan file Tugas1_[NPM]_DLM.res
pt.io.write_results(res_DLM_TOP10, "../Tugas1_DLM.res")

## Evaluation

In [ ]:
# TODO
# import qrels dan tampilkan. (hint: parse qrels dari local test collection)
qrels14 = pt.io.read_qrels("../TUGAS_1/qrels14")
qrels14

qid    docno  label
0     6835     6835      2
1     6835  1440485      1
2     6835  2184490      1
3     6835  1699823      1
4     6835   375405      1
..     ...      ...    ...
777  13690  2212685      1
778  13690  1191518      1
779  13690  1121532      1
780  13690  2163046      1
781  13690  2160218      1

[782 rows x 3 columns]

### BM25

In [ ]:
# TODO
# Melakukan evaluasi hasil retrieval BM25 seluruh query 
# dengan metric evaluasi precision@10, recall@10, dan MRR
from pyterrier.measures import *
eval_BM25_all = pt.Experiment([res_BM25],
                              res_BM25,
                              qrels14,
                              names=["BM25"],
                              eval_metrics=["P_10", "recall_10", RR])
eval_BM25_all

name      P_10  recall_10        RR
0  BM25  0.256667   0.229187  0.662089

In [ ]:
# TODO
# Melakukan evaluasi hasil retrieval BM25 per query 
# dengan metric evaluasi precision@10, recall@10, dan MRR
eval_BM25_per_query = pt.Experiment([res_BM25],
                                    res_BM25,
                                    qrels14,
                                    names=["BM25"],
                                    eval_metrics=["P_10", "recall_10", RR],
                                    perquery=True)
eval_BM25_per_query

name     qid    measure     value
36  BM25  100963         RR  1.000000
37  BM25  100963       P_10  0.500000
38  BM25  100963  recall_10  0.192308
51  BM25  104339         RR  1.000000
52  BM25  104339       P_10  0.600000
..   ...     ...        ...       ...
67  BM25   76048       P_10  0.000000
68  BM25   76048  recall_10  0.000000
45  BM25   85285         RR  0.040000
46  BM25   85285       P_10  0.000000
47  BM25   85285  recall_10  0.000000

[90 rows x 4 columns]

In [ ]:
eval_BM25_per_query_pivot = eval_BM25_per_query.pivot(index=["name","qid"], 
                                                      columns="measure",
                                                      values="value")
eval_BM25_per_query_pivot

measure       P_10        RR  recall_10
name qid                               
BM25 100963    0.5  1.000000   0.192308
     104339    0.6  1.000000   0.545455
     13460     0.2  0.166667   0.250000
     13690     0.2  0.333333   0.333333
     14445     0.5  1.000000   0.142857
     156457    0.3  1.000000   0.375000
     164182    0.1  1.000000   0.111111
     172877    0.2  1.000000   0.222222
     1742574   0.0  0.090909   0.000000
     183397    0.2  1.000000   0.250000
     18363     0.3  0.200000   0.019108
     19145     0.1  0.250000   0.062500
     19342     0.2  1.000000   0.200000
     2228      0.4  1.000000   0.333333
     226344    0.2  1.000000   0.142857
     2613      0.3  1.000000   0.032258
     2674      0.1  1.000000   0.023256
     28296     0.1  0.333333   0.050000
     284       0.3  1.000000   0.272727
     30066     0.2  1.000000   0.200000
     34937     0.2  1.000000   0.333333
     37955     0.4  0.250000   0.307692
     4925      0.3  0.500000   0.015544
     561358    0.7  0.500000   1.000000
     605066    0.3  0.500000   0.500000
     654239    0.3  1.000000   0.375000
     6835      0.3  0.142857   0.300000
     7169      0.2  0.500000   0.285714
     76048     0.0  0.055556   0.000000
     85285     0.0  0.040000   0.000000

In [ ]:
# TODO
# simpan hasil evaluasi dalam bentuk csv dengan row setiap query id 
# dan kolom setiap metrik evaluasi.
# Format penamaan file: Tugas1_[NPM]_eval_BM25.csv
eval_BM25_per_query_pivot.to_csv('../Tugas1_eval_BM25.csv')

### Dirichlet Language Model

In [ ]:
# TODO
# Melakukan evaluasi hasil retrieval Dirichlet Language Model seluruh query dengan metric evaluasi precision@10, recall@10, dan MRR
from pyterrier.measures import *
eval_DLM_all = pt.Experiment([res_DLM],
                             res_DLM,
                             qrels14,
                             names=["DLM"],
                             eval_metrics=["P_10", "recall_10", RR])
eval_DLM_all

name      P_10  recall_10        RR
0  DLM  0.256667   0.223804  0.684121

In [ ]:
# TODO
# Melakukan evaluasi hasil retrieval Dirichlet Language Model per query dengan metrik evaluasi precision@10, recall@10, dan MRR
eval_DLM_per_query = pt.Experiment([res_DLM],
                                   res_DLM,
                                   qrels14,
                                   names=["DLM"],
                                   eval_metrics=["P_10", "recall_10", RR],
                                   perquery=True)
eval_DLM_per_query

name     qid    measure     value
36  DLM  100963         RR  1.000000
37  DLM  100963       P_10  0.500000
38  DLM  100963  recall_10  0.192308
51  DLM  104339         RR  1.000000
52  DLM  104339       P_10  0.600000
..  ...     ...        ...       ...
67  DLM   76048       P_10  0.000000
68  DLM   76048  recall_10  0.000000
45  DLM   85285         RR  0.010000
46  DLM   85285       P_10  0.000000
47  DLM   85285  recall_10  0.000000

[90 rows x 4 columns]

In [ ]:
eval_DLM_per_query_pivot = eval_DLM_per_query.pivot(index=["name","qid"], 
                                                    columns="measure",
                                                    values="value")
eval_DLM_per_query_pivot

measure       P_10        RR  recall_10
name qid                               
DLM  100963    0.5  1.000000   0.192308
     104339    0.6  1.000000   0.545455
     13460     0.2  0.166667   0.250000
     13690     0.2  0.250000   0.333333
     14445     0.5  1.000000   0.142857
     156457    0.2  1.000000   0.250000
     164182    0.1  1.000000   0.111111
     172877    0.2  1.000000   0.222222
     1742574   0.0  0.090909   0.000000
     183397    0.2  1.000000   0.250000
     18363     0.4  1.000000   0.025478
     19145     0.1  0.250000   0.062500
     19342     0.2  1.000000   0.200000
     2228      0.4  1.000000   0.333333
     226344    0.2  1.000000   0.142857
     2613      0.3  1.000000   0.032258
     2674      0.1  1.000000   0.023256
     28296     0.1  0.333333   0.050000
     284       0.3  1.000000   0.272727
     30066     0.2  1.000000   0.200000
     34937     0.2  1.000000   0.333333
     37955     0.4  0.250000   0.307692
     4925      0.3  0.500000   0.015544
     561358    0.6  0.500000   0.857143
     605066    0.3  0.500000   0.500000
     654239    0.3  1.000000   0.375000
     6835      0.4  0.166667   0.400000
     7169      0.2  0.500000   0.285714
     76048     0.0  0.006061   0.000000
     85285     0.0  0.010000   0.000000

In [ ]:
# TODO
# simpan hasil evaluasi dalam bentuk csv dengan row setiap query id dan kolom setiap metrik evaluasi. Format penamaan file: Tugas1_[NPM]_eval_DLM.csv
eval_DLM_per_query_pivot.to_csv('../TUGAS_1/Tugas1_eval_DLM.csv')

Analisis Hasil
1. Lakukan evaluasi hasil keseluruhan query. Manakah metode yang memiliki efektivitas lebih baik antara BM25 dan LM? Apakah perbedaan skor yang diperoleh sigifikan secara statistik?
2. Lakukan evaluasi hasil per query. Pada query mana saja BM25 lebih unggul, dan pada query mana saja LM lebih unggul? Berikan analisis Anda mengapa hal ini bisa terjadi. 
3. Query ID berapa yang memiliki nilai evaluasi terbaik dengan metode BM25? Query ID berapa yang memiliki nilai evaluasi terbaik dengan metode LM? Berikan analisis Anda mengapa hal ini bisa terjadi.
4. Query ID berapa yang memiliki nilai evaluasi terburuk dengan metode BM25? Query ID berapa yang memiliki nilai evaluasi terburuk dengan metode LM? Berikan analisis Anda mengapa hal ini bisa terjadi. 

---

[#TODO: KETIKKAN JAWABAN DISINI]

## **Analisis Hasil**

####**1. Lakukan evaluasi hasil keseluruhan query. Manakah metode yang memiliki efektivitas lebih baik antara BM25 dan LM? Apakah perbedaan skor yang diperoleh sinifikan secara statistik?**

In [ ]:
# Mendefinisikan data untuk dilakukan analisis lebih lanjut
data_retrieve_sign = pt.BatchRetrieve(index)
data_sign = data_retrieve_sign.transform(df)
data_sign

qid   docid    docno  rank      score                query
0       6835  175366  2380216     0  15.529234          joan of arc
1       6835   34385  2185113     1  14.591149          joan of arc
2       6835  253566  1561926     2  14.467350          joan of arc
3       6835  123783   734482     3  14.398326          joan of arc
4       6835  341480  1059708     4  14.289948          joan of arc
...      ...     ...      ...   ...        ...                  ...
19148  13690   59129   495749   995   3.831475  university of tulsa
19149  13690   62293  1102294   996   3.831475  university of tulsa
19150  13690   66622   404623   997   3.831475  university of tulsa
19151  13690   84831   512129   998   3.831475  university of tulsa
19152  13690  106653   872315   999   3.831475  university of tulsa

[19153 rows x 6 columns]

In [ ]:
# dataframe unique qid dan query untuk kebutuhan merge data
# dengan tujuan membandingkan hasil evaluasi per-query dari BM25 dan DLM

data_1 = pd.DataFrame(data_sign, 
                      columns=['qid', 
                               'query'])
data_unique = data_1.drop_duplicates(subset = ["qid"])
data_unique = data_unique.reset_index().drop(["index"], axis=1)
data_unique

qid                                  query
0      6835                            joan of arc
1     19342                      chinese americans
2     18363               fox broadcasting company
3    226344                        parramatta eels
4     37955                               duisburg
5     19145                oxford university press
6     14445                                  yukon
7    605066                tenants purchase scheme
8      2228                         cthulhu mythos
9      2613                              cape town
10  1742574                              phoenicia
11    28296                          equestrianism
12   100963                                taunton
13   561358  1947 ceylonese parliamentary election
14    13460                              triathlon
15    85285                   west coast main line
16     4925                                  ghana
17   104339                   manchester metrolink
18   183397                                 darfur
19   156457                jim crockett promotions
20      284                         antoninus pius
21    34937                           isle of dogs
22    76048                             will smith
23   164182                           dewan rakyat
24    30066                        vitoria gasteiz
25   654239                        wiley blackwell
26     2674                     computer animation
27   172877                         michael bolton
28     7169                     kashubian language
29    13690                    university of tulsa

#### **Signifikansi Testing**


* Misal, metode BM25 akan dijadikan baseline dalam uji signifikansi yang akan dilakukan.

In [ ]:
# Significance Testing
# Kita dapat melakukan pengujian signifikansi dengan 
# menentukan indeks transformator mana yang akan dijadikan sebagai baseline
eval_all_sign = pt.Experiment([res_BM25, res_DLM],
                              data_sign,
                              qrels14,
                              names=["BM25", "DLM"],
                              eval_metrics=["P_10", "recall_10", RR],
                              baseline=0,
                              correction='bonferroni')
eval_all_sign

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


name        RR      P_10  recall_10  RR +  RR -  RR p-value  RR reject  \
0  BM25  0.662089  0.256667   0.229187   NaN   NaN         NaN      False   
1   DLM  0.684121  0.256667   0.223804   2.0   3.0    0.421885      False   

   RR p-value corrected  P_10 +  P_10 -  P_10 p-value  P_10 reject  \
0                   NaN     NaN     NaN           NaN        False   
1               0.84377     2.0     2.0           1.0        False   

   P_10 p-value corrected  recall_10 +  recall_10 -  recall_10 p-value  \
0                     NaN          NaN          NaN                NaN   
1                     1.0          2.0          2.0           0.461214   

   recall_10 reject  recall_10 p-value corrected  
0             False                          NaN  
1             False                     0.922428

* Berdasarkan uji signifikansi yang telah dilakukan di antara metode BM25 dan DLM, dapat disimpulkan bahwa: tidak adanya perbedaan skor yang signifikan secara statistik pada semua metrik evaluasi precision@10, recall@10, dan MRR (`P_10`, `recall_10` dan `RR`). Karena p-value dari semua metrik evaluasi > 0.05.

* Berdasarkan output uji signifikansi tersebut, juga dapat disimpulkan bahwa **metode DLM** meningkatkan nilai presisi_10 rata-rata untuk 2 kueri, meningkatkan nilai recall_10 rata-rata untuk 2 kueri dan juga meningkatkan nilai MRR untuk 2 kueri.

* Berdasarkan output uji signifikansi (secara keseluruhan) tersebut, juga dapat disimpulkan bahwa metode yang memiliki tingkat efektivitas yang lebih baik adalah **metode DLM**.

####**2.Lakukan evaluasi hasil per query. Pada query mana saja BM25 lebih unggul, dan pada query mana saja LM lebih unggul? Berikan analisis Anda mengapa hal ini bisa terjadi**

In [ ]:
# Mendefinisikan kembali hasil evaluasi per-query dari BM25
# dan rename kolom untuk mempermudah proses perbandingan dengan DLM 

eval_BM25_per_query_pivot = eval_BM25_per_query_pivot.copy()
eval_BM25_per_query_pivot
eval_BM25_effect = eval_BM25_per_query_pivot.rename(columns = {'P_10':'P_10_BM25',
                                                              'RR':'RR_BM25',
                                                              'recall_10':'recall_10_BM25'})
eval_BM25_effect

measure       P_10_BM25   RR_BM25  recall_10_BM25
name qid                                         
BM25 100963         0.5  1.000000        0.192308
     104339         0.6  1.000000        0.545455
     13460          0.2  0.166667        0.250000
     13690          0.2  0.333333        0.333333
     14445          0.5  1.000000        0.142857
     156457         0.3  1.000000        0.375000
     164182         0.1  1.000000        0.111111
     172877         0.2  1.000000        0.222222
     1742574        0.0  0.090909        0.000000
     183397         0.2  1.000000        0.250000
     18363          0.3  0.200000        0.019108
     19145          0.1  0.250000        0.062500
     19342          0.2  1.000000        0.200000
     2228           0.4  1.000000        0.333333
     226344         0.2  1.000000        0.142857
     2613           0.3  1.000000        0.032258
     2674           0.1  1.000000        0.023256
     28296          0.1  0.333333        0.050000
     284            0.3  1.000000        0.272727
     30066          0.2  1.000000        0.200000
     34937          0.2  1.000000        0.333333
     37955          0.4  0.250000        0.307692
     4925           0.3  0.500000        0.015544
     561358         0.7  0.500000        1.000000
     605066         0.3  0.500000        0.500000
     654239         0.3  1.000000        0.375000
     6835           0.3  0.142857        0.300000
     7169           0.2  0.500000        0.285714
     76048          0.0  0.055556        0.000000
     85285          0.0  0.040000        0.000000

In [ ]:
# Mendefinisikan kembali hasil evaluasi per-query dari DLM
# dan rename kolom untuk mempermudah proses perbandingan dengan BM25 

eval_DLM_per_query_pivot = eval_DLM_per_query_pivot.copy()
eval_DLM_per_query_pivot
eval_DLM_effect = eval_DLM_per_query_pivot.rename(columns = {'P_10':'P_10_DLM',
                                                              'RR':'RR_DLM',
                                                              'recall_10':'recall_10_DLM'})
eval_DLM_effect

measure       P_10_DLM    RR_DLM  recall_10_DLM
name qid                                       
DLM  100963        0.5  1.000000       0.192308
     104339        0.6  1.000000       0.545455
     13460         0.2  0.166667       0.250000
     13690         0.2  0.250000       0.333333
     14445         0.5  1.000000       0.142857
     156457        0.2  1.000000       0.250000
     164182        0.1  1.000000       0.111111
     172877        0.2  1.000000       0.222222
     1742574       0.0  0.090909       0.000000
     183397        0.2  1.000000       0.250000
     18363         0.4  1.000000       0.025478
     19145         0.1  0.250000       0.062500
     19342         0.2  1.000000       0.200000
     2228          0.4  1.000000       0.333333
     226344        0.2  1.000000       0.142857
     2613          0.3  1.000000       0.032258
     2674          0.1  1.000000       0.023256
     28296         0.1  0.333333       0.050000
     284           0.3  1.000000       0.272727
     30066         0.2  1.000000       0.200000
     34937         0.2  1.000000       0.333333
     37955         0.4  0.250000       0.307692
     4925          0.3  0.500000       0.015544
     561358        0.6  0.500000       0.857143
     605066        0.3  0.500000       0.500000
     654239        0.3  1.000000       0.375000
     6835          0.4  0.166667       0.400000
     7169          0.2  0.500000       0.285714
     76048         0.0  0.006061       0.000000
     85285         0.0  0.010000       0.000000

In [ ]:
# Melakukan merging data antara hasil evaluas perquery dari metode BM25 dan DLM
data_1_merge = pd.merge(data_unique, eval_BM25_effect, on='qid')
data_2_merge = pd.merge(data_1_merge, eval_DLM_effect, on='qid')
data_2_merge

qid                                  query  P_10_BM25   RR_BM25  \
0      6835                            joan of arc        0.3  0.142857   
1     19342                      chinese americans        0.2  1.000000   
2     18363               fox broadcasting company        0.3  0.200000   
3    226344                        parramatta eels        0.2  1.000000   
4     37955                               duisburg        0.4  0.250000   
5     19145                oxford university press        0.1  0.250000   
6     14445                                  yukon        0.5  1.000000   
7    605066                tenants purchase scheme        0.3  0.500000   
8      2228                         cthulhu mythos        0.4  1.000000   
9      2613                              cape town        0.3  1.000000   
10  1742574                              phoenicia        0.0  0.090909   
11    28296                          equestrianism        0.1  0.333333   
12   100963                                taunton        0.5  1.000000   
13   561358  1947 ceylonese parliamentary election        0.7  0.500000   
14    13460                              triathlon        0.2  0.166667   
15    85285                   west coast main line        0.0  0.040000   
16     4925                                  ghana        0.3  0.500000   
17   104339                   manchester metrolink        0.6  1.000000   
18   183397                                 darfur        0.2  1.000000   
19   156457                jim crockett promotions        0.3  1.000000   
20      284                         antoninus pius        0.3  1.000000   
21    34937                           isle of dogs        0.2  1.000000   
22    76048                             will smith        0.0  0.055556   
23   164182                           dewan rakyat        0.1  1.000000   
24    30066                        vitoria gasteiz        0.2  1.000000   
25   654239                        wiley blackwell        0.3  1.000000   
26     2674                     computer animation        0.1  1.000000   
27   172877                         michael bolton        0.2  1.000000   
28     7169                     kashubian language        0.2  0.500000   
29    13690                    university of tulsa        0.2  0.333333   

    recall_10_BM25  P_10_DLM    RR_DLM  recall_10_DLM  
0         0.300000       0.4  0.166667       0.400000  
1         0.200000       0.2  1.000000       0.200000  
2         0.019108       0.4  1.000000       0.025478  
3         0.142857       0.2  1.000000       0.142857  
4         0.307692       0.4  0.250000       0.307692  
5         0.062500       0.1  0.250000       0.062500  
6         0.142857       0.5  1.000000       0.142857  
7         0.500000       0.3  0.500000       0.500000  
8         0.333333       0.4  1.000000       0.333333  
9         0.032258       0.3  1.000000       0.032258  
10        0.000000       0.0  0.090909       0.000000  
11        0.050000       0.1  0.333333       0.050000  
12        0.192308       0.5  1.000000       0.192308  
13        1.000000       0.6  0.500000       0.857143  
14        0.250000       0.2  0.166667       0.250000  
15        0.000000       0.0  0.010000       0.000000  
16        0.015544       0.3  0.500000       0.015544  
17        0.545455       0.6  1.000000       0.545455  
18        0.250000       0.2  1.000000       0.250000  
19        0.375000       0.2  1.000000       0.250000  
20        0.272727       0.3  1.000000       0.272727  
21        0.333333       0.2  1.000000       0.333333  
22        0.000000       0.0  0.006061       0.000000  
23        0.111111       0.1  1.000000       0.111111  
24        0.200000       0.2  1.000000       0.200000  
25        0.375000       0.3  1.000000       0.375000  
26        0.023256       0.1  1.000000       0.023256  
27        0.222222       0.2  1.000000       0.222222  
28        0.285714       0.2  0.500000       0.285714  
29

Berdasarkan proses perhitungan evaluasi hasil dari per-query dapat disimpulkan bahwa:
* Query yang nilai evaluasi BM25-nya lebih unggul dibandingkan dengan DLM adalah
  * 1947 ceylonese parliamentary election	
  * west coast main line
  * jim crockett promotions

* Query yang niai evaluasi DLM-nya lebih unggul dibandingkan dengan BM25 adalah
  * joan of arc
  * fox broadcasting 
  * companywill smith
  * university of tulsa

####**3. Query ID berapa yang memiliki nilai evaluasi terbaik dengan metode BM25? Query ID berapa yang memiliki nilai evaluasi terbaik dengan metode LM? Berikan analisis Anda mengapa hal ini bisa terjadi.**

In [ ]:
# Melakukan sorting (dari value terbesar) untuk memperoleh nilai evaluasi terbaik dari BM25

eval_BM25_per_query_pivot
eval_BM25_per_query_pivot.sort_values(by=['P_10','recall_10','RR'],
                                      ascending=[False,False,False])

measure       P_10        RR  recall_10
name qid                               
BM25 561358    0.7  0.500000   1.000000
     104339    0.6  1.000000   0.545455
     100963    0.5  1.000000   0.192308
     14445     0.5  1.000000   0.142857
     2228      0.4  1.000000   0.333333
     37955     0.4  0.250000   0.307692
     605066    0.3  0.500000   0.500000
     156457    0.3  1.000000   0.375000
     654239    0.3  1.000000   0.375000
     6835      0.3  0.142857   0.300000
     284       0.3  1.000000   0.272727
     2613      0.3  1.000000   0.032258
     18363     0.3  0.200000   0.019108
     4925      0.3  0.500000   0.015544
     34937     0.2  1.000000   0.333333
     13690     0.2  0.333333   0.333333
     7169      0.2  0.500000   0.285714
     183397    0.2  1.000000   0.250000
     13460     0.2  0.166667   0.250000
     172877    0.2  1.000000   0.222222
     19342     0.2  1.000000   0.200000
     30066     0.2  1.000000   0.200000
     226344    0.2  1.000000   0.142857
     164182    0.1  1.000000   0.111111
     19145     0.1  0.250000   0.062500
     28296     0.1  0.333333   0.050000
     2674      0.1  1.000000   0.023256
     1742574   0.0  0.090909   0.000000
     76048     0.0  0.055556   0.000000
     85285     0.0  0.040000   0.000000

Untuk metode BM25, queryID yang memiliki nilai evaluasi terbaik (jika dilihat berdasarkan nilai precision dan recall) adalah:
* qid : 561358 dengan query : 1947 ceylonese parliamentary election.	

In [ ]:
# Melakukan sorting (dari value terbesar) untuk memperoleh nilai evaluasi terbaik dari DLM

eval_DLM_per_query_pivot
eval_DLM_per_query_pivot.sort_values(by=['P_10','recall_10','RR'],
                                     ascending=[False,False,False])

measure       P_10        RR  recall_10
name qid                               
DLM  561358    0.6  0.500000   0.857143
     104339    0.6  1.000000   0.545455
     100963    0.5  1.000000   0.192308
     14445     0.5  1.000000   0.142857
     6835      0.4  0.166667   0.400000
     2228      0.4  1.000000   0.333333
     37955     0.4  0.250000   0.307692
     18363     0.4  1.000000   0.025478
     605066    0.3  0.500000   0.500000
     654239    0.3  1.000000   0.375000
     284       0.3  1.000000   0.272727
     2613      0.3  1.000000   0.032258
     4925      0.3  0.500000   0.015544
     34937     0.2  1.000000   0.333333
     13690     0.2  0.250000   0.333333
     7169      0.2  0.500000   0.285714
     156457    0.2  1.000000   0.250000
     183397    0.2  1.000000   0.250000
     13460     0.2  0.166667   0.250000
     172877    0.2  1.000000   0.222222
     19342     0.2  1.000000   0.200000
     30066     0.2  1.000000   0.200000
     226344    0.2  1.000000   0.142857
     164182    0.1  1.000000   0.111111
     19145     0.1  0.250000   0.062500
     28296     0.1  0.333333   0.050000
     2674      0.1  1.000000   0.023256
     1742574   0.0  0.090909   0.000000
     85285     0.0  0.010000   0.000000
     76048     0.0  0.006061   0.000000

Untuk metode DLM, queryID yang memiliki nilai evaluasi terbaik (jika dilihat berdasarkan nilai precision dan recall) adalah:
* qid : 561358 dengan query : 1947 ceylonese parliamentary election.	

####**4. Query ID berapa yang memiliki nilai evaluasi terburuk dengan metode BM25? Query ID berapa yang memiliki nilai evaluasi terburuk dengan metode LM? Berikan analisis Anda mengapa hal ini bisa terjadi.**

In [ ]:
# Melakukan sorting (dari value terkecil) untuk memperoleh nilai evaluasi terburuk dari BM25

eval_BM25_per_query_pivot
eval_BM25_per_query_pivot.sort_values(by=['P_10','recall_10','RR'],
                                      ascending=[True,True,True])

measure       P_10        RR  recall_10
name qid                               
BM25 85285     0.0  0.040000   0.000000
     76048     0.0  0.055556   0.000000
     1742574   0.0  0.090909   0.000000
     2674      0.1  1.000000   0.023256
     28296     0.1  0.333333   0.050000
     19145     0.1  0.250000   0.062500
     164182    0.1  1.000000   0.111111
     226344    0.2  1.000000   0.142857
     19342     0.2  1.000000   0.200000
     30066     0.2  1.000000   0.200000
     172877    0.2  1.000000   0.222222
     13460     0.2  0.166667   0.250000
     183397    0.2  1.000000   0.250000
     7169      0.2  0.500000   0.285714
     13690     0.2  0.333333   0.333333
     34937     0.2  1.000000   0.333333
     4925      0.3  0.500000   0.015544
     18363     0.3  0.200000   0.019108
     2613      0.3  1.000000   0.032258
     284       0.3  1.000000   0.272727
     6835      0.3  0.142857   0.300000
     156457    0.3  1.000000   0.375000
     654239    0.3  1.000000   0.375000
     605066    0.3  0.500000   0.500000
     37955     0.4  0.250000   0.307692
     2228      0.4  1.000000   0.333333
     14445     0.5  1.000000   0.142857
     100963    0.5  1.000000   0.192308
     104339    0.6  1.000000   0.545455
     561358    0.7  0.500000   1.000000

Untuk metode BM25, queryID yang memiliki nilai evaluasi terburuk (jika dilihat berdasarkan nilai precision dan recall) adalah:
* qid : 85285 dengan query : west coast main line	

In [ ]:
# Melakukan sorting (dari value terkecil) untuk memperoleh nilai evaluasi terburuk dari DLM

eval_DLM_per_query_pivot
eval_DLM_per_query_pivot.sort_values(by=['P_10','recall_10','RR'],
                                     ascending=[True,True,True])

measure       P_10        RR  recall_10
name qid                               
DLM  76048     0.0  0.006061   0.000000
     85285     0.0  0.010000   0.000000
     1742574   0.0  0.090909   0.000000
     2674      0.1  1.000000   0.023256
     28296     0.1  0.333333   0.050000
     19145     0.1  0.250000   0.062500
     164182    0.1  1.000000   0.111111
     226344    0.2  1.000000   0.142857
     19342     0.2  1.000000   0.200000
     30066     0.2  1.000000   0.200000
     172877    0.2  1.000000   0.222222
     13460     0.2  0.166667   0.250000
     156457    0.2  1.000000   0.250000
     183397    0.2  1.000000   0.250000
     7169      0.2  0.500000   0.285714
     13690     0.2  0.250000   0.333333
     34937     0.2  1.000000   0.333333
     4925      0.3  0.500000   0.015544
     2613      0.3  1.000000   0.032258
     284       0.3  1.000000   0.272727
     654239    0.3  1.000000   0.375000
     605066    0.3  0.500000   0.500000
     18363     0.4  1.000000   0.025478
     37955     0.4  0.250000   0.307692
     2228      0.4  1.000000   0.333333
     6835      0.4  0.166667   0.400000
     14445     0.5  1.000000   0.142857
     100963    0.5  1.000000   0.192308
     104339    0.6  1.000000   0.545455
     561358    0.6  0.500000   0.857143

Untuk metode DLM, queryID yang memiliki nilai evaluasi terburuk (jika dilihat berdasarkan nilai precision dan recall) adalah:
* qid : 76048 dengan query : will smith

### Referensi

* https://pyterrier.readthedocs.io/_/downloads/en/latest/pdf/
* https://nlp.stanford.edu/IR-book/essir2011/pdf/12model.pdf
* Bennett, Graham & Scholer, Falk & Uitdenbogerd, Alexandra. (2008). A Comparative Study of Probabalistic and Language Models for Information Retrieval. 65-74.